plugin/trino-hive/src/main/java/io/trino/plugin/hive/fs/TransactionScopeCachingDirectoryLister.java

```diff
@@ -0,0 +1,230 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.hive.fs;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.cache.Cache;
+import io.trino.collect.cache.EvictableCacheBuilder;
+import io.trino.plugin.hive.metastore.Partition;
+import io.trino.plugin.hive.metastore.Storage;
+import io.trino.plugin.hive.metastore.Table;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.LocatedFileStatus;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.RemoteIterator;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Throwables.throwIfInstanceOf;
+import static com.google.common.base.Throwables.throwIfUnchecked;
+import static java.util.Collections.synchronizedList;
+import static java.util.Objects.requireNonNull;
+import static org.apache.commons.lang3.StringUtils.isNotEmpty;
+
+public class TransactionScopeCachingDirectoryLister
+        implements DirectoryLister
+{
+    //TODO use a cache key based on Path & SchemaTableName and iterate over the cache keys
+    // to deal more efficiently with cache invalidation scenarios for partitioned tables.
+    private final Cache<Path, FetchingValueHolder> cache;
+    private final DirectoryLister delegate;
+
+    public TransactionScopeCachingDirectoryLister(DirectoryLister delegate, long maxFileStatuses)
+    {
+        EvictableCacheBuilder<Path, FetchingValueHolder> cacheBuilder = EvictableCacheBuilder.newBuilder()
+                .maximumWeight(maxFileStatuses)
+                .weigher((key, value) -> value.getCachedFilesSize());
+        this.cache = cacheBuilder.build();
+        this.delegate = requireNonNull(delegate, "delegate is null");
+    }
+
+    @Override
+    public RemoteIterator<LocatedFileStatus> list(FileSystem fs, Table table, Path path)
+            throws IOException
+    {
+        FetchingValueHolder cachedValueHolder;
+        try {
+            cachedValueHolder = cache.get(path, () -> new FetchingValueHolder(delegate.list(fs, table, path)));
+        }
+        catch (ExecutionException e) {
+            Throwable throwable = e.getCause();
+            throwIfInstanceOf(throwable, IOException.class);
+            throwIfUnchecked(throwable);
+            throw new RuntimeException("Unexpected cause from delegate DirectoryLister", throwable);
+        }
+
+        if (cachedValueHolder.isFullyCached()) {
+            return cachedValueHolder.getCachedFiles();
+        }
+
+        return cachingRemoteIterator(cachedValueHolder, path);
```

Comment:
if `cachedValueHolder` contains 100 cached entries and then an exception, 
we will create an iterator that will return 100 file statuses and then fail
we could fail fast in such case

what about removing `&& exception == null` from `isFullyCached`?